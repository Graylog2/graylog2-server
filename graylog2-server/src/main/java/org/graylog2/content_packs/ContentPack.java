/**
 * This file is part of Graylog.
 *
 * Graylog is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Graylog is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Graylog.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.graylog2.content_packs;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import org.mongojack.Id;
import org.mongojack.ObjectId;

import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotNull;
import java.util.Collections;
import java.util.Set;

@JsonAutoDetect
@JsonInclude(JsonInclude.Include.NON_NULL)
public class ContentPack {
    @Id
    @ObjectId
    @JsonProperty
    @NotBlank
    private String id;
    @JsonProperty
    @NotBlank
    private String name;
    @JsonProperty
    @NotBlank
    private String version;
    @JsonProperty
    private String description;
    @JsonProperty
    private String summary;
    @JsonProperty
    @NotNull
    private Set<String> states = Collections.emptySet();

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getVersion() {
        return version;
    }

    public void setVersion(String version) {
        this.version = version;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getSummary() {
        return summary;
    }

    public void setSummary(String summary) {
        this.summary = summary;
    }

    public Set<String> getStates() {
        return states;
    }

    public void setStates(Set<String> states) {
        this.states = states;
    }
}
