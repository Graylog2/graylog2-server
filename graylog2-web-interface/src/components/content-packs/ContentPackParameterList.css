:local(.scrollable) tbody {
    max-height: 300px;
    overflow: hidden;
    overflow-y: auto;
    display:block;
}

:local(.scrollable) thead, tbody tr {
    display:table;
    width:100%;
    table-layout:fixed;
}
