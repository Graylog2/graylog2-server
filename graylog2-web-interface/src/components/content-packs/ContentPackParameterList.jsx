import PropTypes from 'prop-types';
import React from 'react';
import { findIndex } from 'lodash';

import { Button, Modal, ButtonToolbar, Badge } from 'react-bootstrap';
import BootstrapModalWrapper from 'components/bootstrap/BootstrapModalWrapper';
import { DataTable, SearchForm } from 'components/common';

import ObjectUtils from 'util/ObjectUtils';
import ContentPackEditParameter from 'components/content-packs/ContentPackEditParameter';

import ContentPackParameterListStyle from './ContentPackParameterList.css';
import ContentPackUtils from './ContentPackUtils';

class ContentPackParameterList extends React.Component {
  static propTypes = {
    contentPack: PropTypes.object.isRequired,
    readOnly: PropTypes.bool,
    onDeleteParameter: PropTypes.func,
    onAddParameter: PropTypes.func,
    appliedParameter: PropTypes.object,
  };

  static defaultProps = {
    readOnly: false,
    onDeleteParameter: () => {},
    onAddParameter: () => {},
    appliedParameter: {},
  };

  constructor(props) {
    super(props);

    this.state = {
      filteredParameters: props.contentPack.parameters || [],
      filter: undefined,
    };
  }

  componentWillReceiveProps(newProps) {
    this._filterParameters(this.state.filter, newProps.contentPack.parameters);
  }

  _parameterApplied = (paramName) => {
    const entityIds = Object.keys(this.props.appliedParameter);
    /* eslint-disable-next-line no-restricted-syntax, guard-for-in */
    for (const i in entityIds) {
      const params = this.props.appliedParameter[entityIds[i]];
      if (findIndex(params, { paramName: paramName }) >= 0) {
        return true;
      }
    }
    return false;
  };

  _parameterRowFormatter = (parameter) => {
    const parameterApplied = this._parameterApplied(parameter.name);
    const buttonTitle = parameterApplied ? 'Still in use' : 'Delete Parameter';
    const icon = parameterApplied ? 'fa fa-check' : 'fa fa-times';
    const bsStyle = parameterApplied ? 'success' : 'failure';
    return (
      <tr key={parameter.title}>
        <td className={ContentPackParameterListStyle.bigColumns}>{parameter.title}</td>
        <td>{parameter.name}</td>
        <td className={ContentPackParameterListStyle.bigColumns}>{parameter.description}</td>
        <td>{parameter.type}</td>
        <td>{ContentPackUtils.convertToString(parameter)}</td>
        <td><Badge className={bsStyle}><i className={icon} /></Badge></td>
        {!this.props.readOnly &&
        <td>
          <ButtonToolbar>
            <Button bsStyle="primary"
                    bsSize="xs"
                    title={buttonTitle}
                    disabled={parameterApplied}
                    onClick={() => { this.props.onDeleteParameter(parameter); }}>
              Delete
            </Button>{this._parameterModal(parameter)}
          </ButtonToolbar>
        </td>
        }
      </tr>
    );
  };

  _parameterModal(parameter) {
    let modalRef;
    let editParameter;

    const closeModal = () => {
      modalRef.close();
    };

    const openModal = () => {
      modalRef.open();
    };

    const addParameter = () => {
      editParameter.addNewParameter();
    };

    const size = parameter ? 'xsmall' : 'small';
    const name = parameter ? 'Edit' : 'Create parameter';

    const modal = (
      <BootstrapModalWrapper ref={(node) => { modalRef = node; }} bsSize="large">
        <Modal.Header closeButton>
          <Modal.Title>Parameter</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <ContentPackEditParameter ref={(node) => { editParameter = node; }}
                                    parameters={this.props.contentPack.parameters}
                                    onUpdateParameter={(newParameter) => {
                                      this.props.onAddParameter(newParameter, parameter);
                                      closeModal();
                                    }}
                                    parameterToEdit={parameter} />
        </Modal.Body>
        <Modal.Footer>
          <div className="pull-right">
            <ButtonToolbar>
              <Button bsStyle="primary" onClick={addParameter}>Save</Button>
              <Button onClick={closeModal}>Close</Button>
            </ButtonToolbar>
          </div>
        </Modal.Footer>
      </BootstrapModalWrapper>
    );

    return (
      <React.Fragment>
        <Button bsStyle="info"
                bsSize={size}
                title="Edit Modal"
                onClick={openModal}>
          {name}
        </Button>
        {modal}
      </React.Fragment>
    );
  }

  _filterParameters = (filter, parametersArg) => {
    const parameters = ObjectUtils.clone(parametersArg || this.props.contentPack.parameters);
    if (!filter || filter.length <= 0) {
      this.setState({ filteredParameters: parameters, filter: undefined });
      return;
    }
    const regexp = RegExp(filter, 'i');
    const filteredParameters = parameters.filter((parameter) => {
      return regexp.test(parameter.title) || regexp.test(parameter.description) || regexp.test(parameter.name);
    });

    this.setState({ filteredParameters: filteredParameters, filter: filter });
  };

  render() {
    const headers = this.props.readOnly ?
      ['Title', 'Name', 'Description', 'Value Type', 'Default Value', 'Used'] :
      ['Title', 'Name', 'Description', 'Value Type', 'Default Value', 'Used', 'Action'];
    return (
      <div>
        <h2>Parameters list</h2>
        <br />
        { !this.props.readOnly && this._parameterModal() }
        { !this.props.readOnly && (<span><br /><br /></span>) }
        <SearchForm
          onSearch={this._filterParameters}
          onReset={() => { this._filterParameters(''); }}
          searchButtonLabel="Filter"
        />
        <DataTable
          id="parameter-list"
          headers={headers}
          className={ContentPackParameterListStyle.scrollable}
          sortByKey="title"
          noDataText="To use parameters for content packs, at first a parameter must be created and can then be applied to a entity."
          filterKeys={[]}
          rows={this.state.filteredParameters}
          dataRowFormatter={this._parameterRowFormatter}
        />
      </div>
    );
  }
}

export default ContentPackParameterList;
