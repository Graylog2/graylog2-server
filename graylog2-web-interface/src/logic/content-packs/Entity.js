import { Map } from 'immutable';
import { findIndex } from 'lodash';
import ValueRefHelper from 'util/ValueRefHelper';
import Constraint from './Constraint';

export default class Entity {
  constructor(v, type, id, data, fromServer = false, constraintValues = [], parameters = []) {
    const constraints = constraintValues.map((c) => {
      if (c instanceof Constraint) {
        return c;
      }
      return Constraint.fromJSON(c);
    });

    this._value = {
      v,
      type,
      id,
      data,
      constraints,
      fromServer,
      parameters,
    };
  }

  static fromJSON(value, fromServer = true, parameters = []) {
    const { v, type, id, data, constraints } = value;
    return new Entity(v, type, id, data, fromServer, constraints, parameters);
  }

  get v() {
    return this._value.v;
  }

  get type() {
    return this._value.type;
  }

  get id() {
    return this._value.id;
  }

  get data() {
    return this._value.data;
  }

  get fromServer() {
    return this._value.fromServer;
  }

  get constraints() {
    return this._value.constraints;
  }

  get title() {
    let value = this.getValueFromData('title');
    if (!value) {
      value = this.getValueFromData('name');
    }
    return value || '';
  }

  get description() {
    return this.getValueFromData('description') || '';
  }

  /* eslint-disable-next-line class-methods-use-this */
  get isEntity() {
    return true;
  }

  /* implement custom instanceof */
  static [Symbol.hasInstance](obj) {
    if (obj.isEntity) {
      return true;
    }
    return false;
  }

  getValueFromData(key) {
    const { data } = this._value;
    if (!data || !data[key]) {
      return undefined;
    }

    if (ValueRefHelper.dataIsValueRef(data[key])) {
      const value = (data[key] || {})[ValueRefHelper.VALUE_REF_VALUE_FIELD];
      if (ValueRefHelper.dataValueIsParameter(data[key])) {
        const index = findIndex(this._value.parameters, { name: value });
        if (index >= 0 && this._value.parameters[index].default_value) {
          return this._value.parameters[index].default_value;
        }
      }
      return value;
    }
    return data[key];
  }

  toBuilder() {
    const {
      v,
      type,
      id,
      data,
      constraints,
      fromServer,
      parameters,
    } = this._value;
    /* eslint-disable-next-line no-use-before-define */
    return new Builder(Map({
      v,
      type,
      id,
      data,
      constraints,
      fromServer,
      parameters,
    }));
  }


  static builder() {
    /* eslint-disable-next-line no-use-before-define */
    return new Builder();
  }

  toJSON() {
    const {
      v,
      type,
      id,
      data,
      constraints,
    } = this._value;
    return {
      v,
      type,
      id,
      data,
      constraints,
    };
  }
}

class Builder {
  constructor(value = Map()) {
    this.value = value;
  }

  v(value) {
    this.value = this.value.set('v', value);
    return this;
  }

  type(value) {
    this.value = this.value.set('type', value);
    return this;
  }

  id(value) {
    this.value = this.value.set('id', value);
    return this;
  }

  data(value) {
    this.value = this.value.set('data', value);
    return this;
  }

  fromServer(value) {
    this.value = this.value.set('fromServer', value);
    return this;
  }

  constraints(value) {
    this.value = this.value.set('constraints', value);
    return this;
  }

  parameters(value) {
    this.value = this.value.set('parameters', value);
    return this;
  }

  build() {
    const {
      v,
      type,
      id,
      data,
      constraints,
      fromServer,
      parameters,
    } = this.value.toObject();
    return new Entity(v, type, id, data, fromServer, constraints, parameters);
  }
}
