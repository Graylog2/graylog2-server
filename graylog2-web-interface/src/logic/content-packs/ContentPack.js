import { Map, Set } from 'immutable';
import { concat, remove } from 'lodash';
import uuid from 'uuid/v4';
import Entity from './Entity';

export default class ContentPack {
  constructor(v, id, rev, name, summary, description, vendor, url, parameters, entitieValues) {
    const entities = entitieValues.map((e) => {
      if (e instanceof Entity) {
        return e;
      }
      return Entity.fromJSON(e, false, parameters);
    });

    this._value = {
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    };
  }

  get v() {
    return this._value.v;
  }

  get id() {
    return this._value.id;
  }

  get rev() {
    return this._value.rev;
  }

  get name() {
    return this._value.name;
  }

  get summary() {
    return this._value.summary;
  }

  get description() {
    return this._value.description;
  }

  get vendor() {
    return this._value.vendor;
  }

  get url() {
    return this._value.url;
  }

  get constraints() {
    return this._value.entities.reduce((acc, entity) => {
      return entity.constraints.reduce((result, constraint) => {
        return result.add(constraint);
      }, acc);
    }, Set());
  }

  get parameters() {
    return this._value.parameters;
  }

  get entities() {
    return this._value.entities;
  }

  toBuilder() {
    const {
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    } = this._value;
    // eslint-disable-next-line no-use-before-define
    return new Builder(Map({
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    }));
  }

  toJSON() {
    const {
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    } = this._value;

    const entitiesJSON = entities.map(e => e.toJSON());
    return {
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities: entitiesJSON,
    };
  }

  static fromJSON(value) {
    const {
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    } = value;
    return new ContentPack(
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    );
  }

  static builder() {
    // eslint-disable-next-line no-use-before-define
    return new Builder()
      .v(1)
      .id(uuid())
      .rev(1)
      .name('')
      .summary('')
      .description('')
      .vendor('')
      .url('')
      .parameters([])
      .entities([]);
  }
}

class Builder {
  constructor(value = Map()) {
    this.value = value;
  }

  v(value) {
    this.value = this.value.set('v', value);
    return this;
  }

  id(value) {
    this.value = this.value.set('id', value);
    return this;
  }

  rev(value) {
    this.value = this.value.set('rev', value);
    return this;
  }

  name(value) {
    this.value = this.value.set('name', value);
    return this;
  }

  summary(value) {
    this.value = this.value.set('summary', value);
    return this;
  }

  description(value) {
    this.value = this.value.set('description', value);
    return this;
  }

  vendor(value) {
    this.value = this.value.set('vendor', value);
    return this;
  }

  url(value) {
    this.value = this.value.set('url', value);
    return this;
  }

  parameters(value) {
    this.value = this.value.set('parameters', value);
    return this;
  }

  removeParameter(value) {
    const parameters = this.value.get('parameters').slice(0);
    remove(parameters, parameter => parameter.name === value.name);
    this.value = this.value.set('parameters', parameters);
    return this;
  }

  addParameter(value) {
    const parameters = this.value.get('parameters');
    const newParameters = concat(parameters, value);
    this.value = this.value.set('parameters', newParameters);
    return this;
  }

  entities(value) {
    this.value = this.value.set('entities', value);
    return this;
  }

  build() {
    const {
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    } = this.value.toObject();
    return new ContentPack(
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    );
  }
}
