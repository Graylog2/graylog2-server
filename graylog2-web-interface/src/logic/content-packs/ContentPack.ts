/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import { Map, Set } from 'immutable';
import concat from 'lodash/concat';
import remove from 'lodash/remove';

import generateId from 'logic/generateId';
import type { ContentPackEntity, ContentPackParameter } from 'components/content-packs/Types';

import Entity from './Entity';

export default class ContentPack {
  private _value: {
    v: number;
    id: string;
    rev: number;
    name: string;
    summary: string;
    description: string;
    vendor: string;
    url: string;
    parameters: ContentPackParameter[];
    entities: ContentPackEntity[];
  };

  constructor(
    v: number,
    id: string,
    rev: number,
    name: string,
    summary: string,
    description: string,
    vendor: string,
    url: string,
    parameters: ContentPackParameter[],
    entitieValues: ContentPackEntity[],
  ) {
    const entities = entitieValues.map((e) => {
      if (e instanceof Entity) {
        return e;
      }

      return Entity.fromJSON(e, false, parameters);
    });

    this._value = {
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities,
    };
  }

  get v() {
    return this._value.v;
  }

  get id() {
    return this._value.id;
  }

  get rev() {
    return this._value.rev;
  }

  get name() {
    return this._value.name;
  }

  get summary() {
    return this._value.summary;
  }

  get description() {
    return this._value.description;
  }

  get vendor() {
    return this._value.vendor;
  }

  get url() {
    return this._value.url;
  }

  get constraints() {
    return this._value.entities.reduce(
      (acc, entity) => entity.constraints.reduce((result, constraint) => result.add(constraint), acc),
      Set(),
    );
  }

  get parameters() {
    return this._value.parameters;
  }

  get entities() {
    return this._value.entities;
  }

  toBuilder() {
    const { v, id, rev, name, summary, description, vendor, url, parameters, entities } = this._value;

    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return new Builder(
      Map({
        v,
        id,
        rev,
        name,
        summary,
        description,
        vendor,
        url,
        parameters,
        entities,
      }),
    );
  }

  toJSON() {
    const { v, id, rev, name, summary, description, vendor, url, parameters, entities } = this._value;

    const entitiesJSON = entities.map((e) => e.toJSON());

    return {
      v,
      id,
      rev,
      name,
      summary,
      description,
      vendor,
      url,
      parameters,
      entities: entitiesJSON,
    };
  }

  static fromJSON(value) {
    const { v, id, rev, name, summary, description, vendor, url, parameters, entities } = value;

    return new ContentPack(v, id, rev, name, summary, description, vendor, url, parameters, entities);
  }

  static builder() {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return new Builder()
      .v(1)
      .id(generateId())
      .rev(1)
      .name('')
      .summary('')
      .description('')
      .vendor('')
      .url('')
      .parameters([])
      .entities([]);
  }
}

class Builder {
  private value: Map<string, unknown>;

  constructor(value: Map<string, unknown> = Map()) {
    this.value = value;
  }

  v(value: number) {
    this.value = this.value.set('v', value);

    return this;
  }

  id(value: string) {
    this.value = this.value.set('id', value);

    return this;
  }

  rev(value: number) {
    this.value = this.value.set('rev', value);

    return this;
  }

  name(value: string) {
    this.value = this.value.set('name', value);

    return this;
  }

  summary(value: string) {
    this.value = this.value.set('summary', value);

    return this;
  }

  description(value: string) {
    this.value = this.value.set('description', value);

    return this;
  }

  vendor(value: string) {
    this.value = this.value.set('vendor', value);

    return this;
  }

  url(value: string) {
    this.value = this.value.set('url', value);

    return this;
  }

  parameters(value: ContentPackParameter[]) {
    this.value = this.value.set('parameters', value);

    return this;
  }

  private getParameters() {
    return this.value.get('parameters') as ContentPackParameter[];
  }

  removeParameter(value: ContentPackParameter) {
    const parameters = this.getParameters().slice(0);

    remove(parameters, (parameter) => parameter.name === value.name);
    this.value = this.value.set('parameters', parameters);

    return this;
  }

  addParameter(value: ContentPackParameter) {
    const parameters = this.getParameters();
    const newParameters = concat(parameters, value);

    this.value = this.value.set('parameters', newParameters);

    return this;
  }

  entities(value: Array<ContentPackEntity>) {
    this.value = this.value.set('entities', value);

    return this;
  }

  build() {
    const { v, id, rev, name, summary, description, vendor, url, parameters, entities } = this.value.toObject();

    // @ts-ignore
    return new ContentPack(v, id, rev, name, summary, description, vendor, url, parameters, entities);
  }
}
