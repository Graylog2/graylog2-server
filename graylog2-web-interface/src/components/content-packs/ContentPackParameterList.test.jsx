import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import 'helpers/mocking/react-dom_mock';

import ContentPack from 'logic/content-packs/ContentPack';
import Entity from 'logic/content-packs/Entity';
import ContentPackParameterList from 'components/content-packs/ContentPackParameterList';

describe('<ContentPackParameterList />', () => {
  it('should render with empty parameters with readOnly', () => {
    const contentPack = ContentPack.builder().build();
    const wrapper = renderer.create(<ContentPackParameterList contentPack={contentPack} readOnly />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should render with parameters with readOnly', () => {
    const parameters = [{
      name: 'PARAM',
      title: 'A parameter title',
      description: 'A parameter descriptions',
      type: 'string',
      default_value: 'test',
    }];
    const contentPack = ContentPack.builder()
      .parameters(parameters)
      .build();
    const wrapper = renderer.create(<ContentPackParameterList contentPack={contentPack} readOnly />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should render with empty parameters without readOnly', () => {
    const contentPack = ContentPack.builder().build();
    const wrapper = renderer.create(<ContentPackParameterList contentPack={contentPack} />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should render with parameters without readOnly', () => {
    const parameters = [{
      name: 'PARAM',
      title: 'A parameter title',
      description: 'A parameter descriptions',
      type: 'string',
      default_value: 'test',
    }];
    const contentPack = ContentPack.builder()
      .parameters(parameters)
      .build();
    const wrapper = renderer.create(<ContentPackParameterList contentPack={contentPack} />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should delete a parameter', () => {
    const deleteFn = jest.fn((parameter) => {
      expect(parameter.name).toEqual('PARAM');
    });
    const parameters = [{
      name: 'PARAM',
      title: 'A parameter title',
      description: 'A parameter descriptions',
      type: 'string',
      default_value: 'test',
    }];
    const contentPack = ContentPack.builder()
      .parameters(parameters)
      .build();

    const wrapper = mount(<ContentPackParameterList contentPack={contentPack}
                                                    onDeleteParameter={deleteFn} />);
    wrapper.find('button[children="Delete"]').simulate('click');
    expect(deleteFn.mock.calls.length).toBe(1);
  });

  it('should not delete a used parameter', () => {
    const entity = Entity.builder()
      .v(1)
      .type('input')
      .id('dead-beef')
      .data({ title: { '@type': 'parameter', '@value': 'PARAM' } })
      .build();
    const deleteFn = jest.fn((parameter) => {
      expect(parameter.name).toEqual('PARAM');
    });
    const parameters = [{
      name: 'PARAM',
      title: 'A parameter title',
      description: 'A parameter descriptions',
      type: 'string',
      default_value: 'test',
    }];
    const appliedParameter = {
      'dead-beef': [{ paramName: 'PARAM', configKey: 'title' }],
    };
    const contentPack = ContentPack.builder()
      .parameters(parameters)
      .entities([entity])
      .build();

    const wrapper = mount(<ContentPackParameterList contentPack={contentPack}
                                                    onDeleteParameter={deleteFn}
                                                    appliedParameter={appliedParameter} />);
    wrapper.find('button[children="Delete"]').simulate('click');
    expect(deleteFn.mock.calls.length).toBe(0);
  });

  it('should filter parameters', () => {
    const parameters = [{
      name: 'PARAM',
      title: 'A parameter title',
      description: 'A parameter descriptions',
      type: 'string',
      default_value: 'test',
    }, {
      name: 'BAD_PARAM',
      title: 'real bad',
      description: 'The dark ones own parameter',
      type: 'string',
      default_value: 'test',
    }];
    const contentPack = ContentPack.builder()
      .parameters(parameters)
      .build();
    const wrapper = mount(<ContentPackParameterList contentPack={contentPack} />);
    expect(wrapper.find("td[children='PARAM']").exists()).toBe(true);
    wrapper.find('input').simulate('change', { target: { value: 'Bad' } });
    wrapper.find('form').simulate('submit');
    expect(wrapper.find("td[children='PARAM']").exists()).toBe(false);
    wrapper.find("button[children='Reset']").simulate('click');
    expect(wrapper.find("td[children='PARAM']").exists()).toBe(true);
  });
});
