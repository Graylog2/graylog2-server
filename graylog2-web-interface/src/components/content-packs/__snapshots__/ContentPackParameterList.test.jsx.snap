// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContentPackParameterList /> should render with empty parameters with readOnly 1`] = `
<div>
  <h2>
    Parameters list
  </h2>
  <br />
  <div
    className="search"
    style={
      Object {
        "marginTop": 15,
      }
    }
  >
    <form
      className="form-inline"
      onSubmit={[Function]}
    >
      <div
        className="form-group has-feedback"
      >
        <input
          autoComplete="off"
          className="query form-control"
          id="common-search-form-query-input"
          onChange={[Function]}
          placeholder="Enter search query..."
          spellCheck="false"
          style={
            Object {
              "width": "auto",
            }
          }
          type="text"
          value=""
        />
      </div>
      <div
        className="form-group"
        style={
          Object {
            "marginLeft": 5,
          }
        }
      >
        <button
          className="submit-button btn btn-default"
          disabled={false}
          type="submit"
        >
          Filter
        </button>
      </div>
      <div
        className="form-group"
        style={
          Object {
            "marginLeft": 5,
          }
        }
      >
        <button
          className="reset-button btn btn-default"
          disabled={false}
          onClick={[Function]}
          type="reset"
        >
          Reset
        </button>
      </div>
    </form>
  </div>
  <div>
    <div
      className="row "
    >
      <div
        className="col-md-12"
      >
        <div
          className="data-table table-responsive"
          id="parameter-list"
        >
          <p>
            To use parameters for content packs, at first a parameter must be created and can then be applied to a entity.
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<ContentPackParameterList /> should render with empty parameters without readOnly 1`] = `
<div>
  <h2>
    Parameters list
  </h2>
  <br />
  <button
    className="btn btn-sm btn-info"
    disabled={false}
    onClick={[Function]}
    title="Edit Modal"
    type="button"
  >
    Create parameter
  </button>
  <span>
    <br />
    <br />
  </span>
  <div
    className="search"
    style={
      Object {
        "marginTop": 15,
      }
    }
  >
    <form
      className="form-inline"
      onSubmit={[Function]}
    >
      <div
        className="form-group has-feedback"
      >
        <input
          autoComplete="off"
          className="query form-control"
          id="common-search-form-query-input"
          onChange={[Function]}
          placeholder="Enter search query..."
          spellCheck="false"
          style={
            Object {
              "width": "auto",
            }
          }
          type="text"
          value=""
        />
      </div>
      <div
        className="form-group"
        style={
          Object {
            "marginLeft": 5,
          }
        }
      >
        <button
          className="submit-button btn btn-default"
          disabled={false}
          type="submit"
        >
          Filter
        </button>
      </div>
      <div
        className="form-group"
        style={
          Object {
            "marginLeft": 5,
          }
        }
      >
        <button
          className="reset-button btn btn-default"
          disabled={false}
          onClick={[Function]}
          type="reset"
        >
          Reset
        </button>
      </div>
    </form>
  </div>
  <div>
    <div
      className="row "
    >
      <div
        className="col-md-12"
      >
        <div
          className="data-table table-responsive"
          id="parameter-list"
        >
          <p>
            To use parameters for content packs, at first a parameter must be created and can then be applied to a entity.
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<ContentPackParameterList /> should render with parameters with readOnly 1`] = `
<div>
  <h2>
    Parameters list
  </h2>
  <br />
  <div
    className="search"
    style={
      Object {
        "marginTop": 15,
      }
    }
  >
    <form
      className="form-inline"
      onSubmit={[Function]}
    >
      <div
        className="form-group has-feedback"
      >
        <input
          autoComplete="off"
          className="query form-control"
          id="common-search-form-query-input"
          onChange={[Function]}
          placeholder="Enter search query..."
          spellCheck="false"
          style={
            Object {
              "width": "auto",
            }
          }
          type="text"
          value=""
        />
      </div>
      <div
        className="form-group"
        style={
          Object {
            "marginLeft": 5,
          }
        }
      >
        <button
          className="submit-button btn btn-default"
          disabled={false}
          type="submit"
        >
          Filter
        </button>
      </div>
      <div
        className="form-group"
        style={
          Object {
            "marginLeft": 5,
          }
        }
      >
        <button
          className="reset-button btn btn-default"
          disabled={false}
          onClick={[Function]}
          type="reset"
        >
          Reset
        </button>
      </div>
    </form>
  </div>
  <div>
    <div
      className="row "
    >
      <div
        className="col-md-12"
      >
        <div
          className="data-table table-responsive"
          id="parameter-list"
        >
          <table
            className="table scrollable"
          >
            <thead>
              <tr>
                <th>
                  Title
                </th>
                <th>
                  Name
                </th>
                <th>
                  Description
                </th>
                <th>
                  Value Type
                </th>
                <th>
                  Default Value
                </th>
                <th>
                  Used
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td
                  className="bigColumns"
                >
                  A parameter title
                </td>
                <td>
                  PARAM
                </td>
                <td
                  className="bigColumns"
                >
                  A parameter descriptions
                </td>
                <td>
                  string
                </td>
                <td>
                  test
                </td>
                <td>
                  <span
                    className="failure badge"
                  >
                    <i
                      className="fa fa-times"
                    />
                  </span>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<ContentPackParameterList /> should render with parameters without readOnly 1`] = `
<div>
  <h2>
    Parameters list
  </h2>
  <br />
  <button
    className="btn btn-sm btn-info"
    disabled={false}
    onClick={[Function]}
    title="Edit Modal"
    type="button"
  >
    Create parameter
  </button>
  <span>
    <br />
    <br />
  </span>
  <div
    className="search"
    style={
      Object {
        "marginTop": 15,
      }
    }
  >
    <form
      className="form-inline"
      onSubmit={[Function]}
    >
      <div
        className="form-group has-feedback"
      >
        <input
          autoComplete="off"
          className="query form-control"
          id="common-search-form-query-input"
          onChange={[Function]}
          placeholder="Enter search query..."
          spellCheck="false"
          style={
            Object {
              "width": "auto",
            }
          }
          type="text"
          value=""
        />
      </div>
      <div
        className="form-group"
        style={
          Object {
            "marginLeft": 5,
          }
        }
      >
        <button
          className="submit-button btn btn-default"
          disabled={false}
          type="submit"
        >
          Filter
        </button>
      </div>
      <div
        className="form-group"
        style={
          Object {
            "marginLeft": 5,
          }
        }
      >
        <button
          className="reset-button btn btn-default"
          disabled={false}
          onClick={[Function]}
          type="reset"
        >
          Reset
        </button>
      </div>
    </form>
  </div>
  <div>
    <div
      className="row "
    >
      <div
        className="col-md-12"
      >
        <div
          className="data-table table-responsive"
          id="parameter-list"
        >
          <table
            className="table scrollable"
          >
            <thead>
              <tr>
                <th>
                  Title
                </th>
                <th>
                  Name
                </th>
                <th>
                  Description
                </th>
                <th>
                  Value Type
                </th>
                <th>
                  Default Value
                </th>
                <th>
                  Used
                </th>
                <th>
                  Action
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td
                  className="bigColumns"
                >
                  A parameter title
                </td>
                <td>
                  PARAM
                </td>
                <td
                  className="bigColumns"
                >
                  A parameter descriptions
                </td>
                <td>
                  string
                </td>
                <td>
                  test
                </td>
                <td>
                  <span
                    className="failure badge"
                  >
                    <i
                      className="fa fa-times"
                    />
                  </span>
                </td>
                <td>
                  <div
                    className="btn-toolbar"
                    role="toolbar"
                  >
                    <button
                      className="btn btn-xs btn-primary"
                      disabled={false}
                      onClick={[Function]}
                      title="Delete Parameter"
                      type="button"
                    >
                      Delete
                    </button>
                    <button
                      className="btn btn-xs btn-info"
                      disabled={false}
                      onClick={[Function]}
                      title="Edit Modal"
                      type="button"
                    >
                      Edit
                    </button>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
